lib/engine/round/g_1870/operating.rb

```diff
@@ -7,13 +7,83 @@ module Engine
   module Round
     module G1870
       class Operating < Operating
-        attr_accessor :river_special_tile_lay
+        attr_accessor :river_special_tile_lay, :connection_steps
+        attr_reader :connection_runs
 
         def start_operating
           super
 
           @river_special_tile_lay = nil
         end
+
+        def setup
+          check_connection_runs
+          @connection_steps = []
+
+          super
+        end
+
+        def next_entity!
+          check_connection_runs
+          return if @connection_runs.any?
+
+          super
+        end
+
+        def active_entities
+          return @connection_runs.keys if @connection_runs.any?
+
+          super
+        end
+
+        def check_connection_runs
+          @connection_runs = {}
+          corporations = @game.corporations.dup.sort
+          if current_entity && corporations.any?(current_entity)
+            corporations.unshift(corporations.delete(current_entity))
+          end
+
+          corporations.each do |corporation|
+            next unless (bound = corporation.trains.map(&:distance).max)
+            next unless (destination = @game.destination_hex(corporation))
+            next unless destination.assigned?(corporation)
+
+            home = @game.home_hex(corporation)
+
+            distance = lambda do |hex|
+              (home.x - hex.x)**2 / 4 + (home.y - hex.y)**2
+            end
+
+            next unless check_connection_run(corporation, distance, bound, home, { destination => 1 }, [])
+
+            corporation.trains.each { |train| train.operated = false }
+
+            @connection_runs[corporation] = destination
+            destination.remove_assignment!(corporation)
+
+            @game.log << "-- #{corporation.name} can connect to its destination --"
+          end
+        end
+
+        def check_connection_run(corporation, distance, bound, home, queue, visited)
+          return false unless (current = queue.keys.min_by(&distance))
```

Comment:
I'm a little concerned that this will be a massive CPU hog.

Could the approach taken by 1860 in compute_distance_graph be generalized?

Reply:
It isn't. The algorithm is O(V+E), with some heuristics. I ran stackprof on a completed game and it only accounted for 12% of the computation.

I have no idea what 1860, because this was written 2 months ago and right now I don't have the time nor bandwidth to figure it out again.